binary/function.proto

```diff
@@ -36,10 +33,7 @@ message FunctionSignature {
         bool deterministic = 7;
         bool session_dependent = 8;
 
-        oneof derivation_type {
-            Type direct = 9;
-            CustomDerivationPolicy derivation_policy = 10;
-        }
+        TypeExpression output_type = 9;
```

Comment:
Any particular reason intermediate_type is a concrete type rather than TypeExpression?

Also, what is max_set? In general, how do you feel about adding comments to describe various fields here to help the reader?